### ISSUE TITLE
libdfp: x86 interoperability with gcc

### ISSUE THREAD

**Issue by kenpocowboy** [libdfp/libdfp#118]
Here is a small C program showing that GCC _Decimal128 and libdfp _Decimal128 don't play well together.

```
#define __STDC_WANT_DEC_FP__ 1
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <decimal128.h>
#include <decNumber.h>

int main (const int argc, const char * const argv[]) {
	_Decimal128 value = 6.0DL;
	decNumber result;
	char string[DECIMAL128_String];
	fprintf(stderr,"_Decimal128: %DDE\n",value);
	decimal128ToNumber((decimal128*)&value, &result);
	decNumberToString(&result, (char*)&string);
	fprintf(stderr,"decNumber:   %s\n",(char*)string);
	return 0;
	}
```
When running the code, the results from decimal128ToNumber are incorrect:

```
64 bit
_Decimal128: 6.000000E+00
decNumber:   1.5920E+1508

32 bit
_Decimal128: 6.000000E+00
decNumber:   1.5920E+1508
```

**Comment by kenpocowboy** [comment]
I really only need support for decimal128ToNumber, but for consistency, decimal128FromNumber should work as well.

Most of my work is done either in _Decimal128, or decNumber.

decNumber is required, not only for higher precision, but also as an intermediary betwixt _Decimal128 and other data types.

**Comment by tuliom** [comment]
This is duplicated with issue #78 .
If this program is linked against a libdecnumber from GCC, it prints the right result:

```
_Decimal128: 6.000000E+00
decNumber:   6.0
```

**Comment by kenpocowboy** [comment]
I, obviously, missed that when perusing the issues - sorry !

**Comment by tuliom** [comment]
That's not a problem.   I plan to leave this issue open because you've just provided a testcase that should be part of one of our testcases.
Thanks!

**Comment by pmur** [comment]
This is not a bug. It is an internal implementation difference between the two implementations, both in how BID is converted and how little endian support was added.

The libdfp _DecimalN <-> decNumber conversions assume the value is in big-endian. libgcc assumes host ordering. I don't think IEEE mandates a specific byte-ordering as libdfp would have you believe.  Using the public libdfp API, this subtle difference should be hidden (e.g libdfp takes care to call __ieee_X_to_host and __host_to_ieee_X as needed). decnumber was effectively forked when BID support was added here.

Likewise, I think explicitly exposing decNumber API may be beneficial, but that discussion should be contained in a separate issue.

**Comment by ryanarn** [comment]
Paul's analysis is correct. When I created libdfp I was working on it
before the ratification of BID into the IEEE 854r DFP spec (based on the
original sources from Mike Cowlishaw). I recall that this implementation
presumed Big-Endian.  After the ratification I didn't go out of my way to
support little-endian or BID compatibility but I did try to leave the
library flexible at build level. I believe the response I got from IBM
architects at the time was "Little Endian on Power? That's sacrilegious...
same response I got to PC relative addressing on RISC."

I never intended decNumber to be forked. I intended for it to be updated in
libdfp as it was updated in GCC. Unfortunately Autoconf, at the time (and
perhaps still now), lacked the features necessary to easily import sources
from an external tree at build time, so I begrudgingly imported them as
suggested by the GCC community.

Imho, if possible, it'd be best to have a single libdecnumber source.
